### PR TITLE
IDEM-193&181

### PR DESCRIPTION
**Pull request by xiangzhe0812**
1. basic auth
2. end point for 193
3. part of 181
Swagger url: http://localhost:3000/swagger/#/

### CODE REVIEW

File: src/services/ThirdPartyService.ts [hunk 1]
```diff
@@ -40,4 +41,36 @@ export class ThirdPartyService implements IThirdPartyService {
...
+
+    public async signup(signupInfo: UserSignupRequest): Promise<string> {
+        const requestBody = JSON.stringify({
+            firstName: signupInfo?.firstName,
+            lastName: signupInfo?.lastName,
+            email: signupInfo.email,
+            password: signupInfo.password,
+            referralCode: "",
+            trackAddress: true,
+            CreateAddress: true
```
Comment (bitcoinbrisbane) [inline src/services/ThirdPartyService.ts:53]:
```suggestion
            createAddress: true
```

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Good start.  

- Should we look at convention for column names?
- Also should we add the vendors to a db now?

File: src/interfaces.ts [hunk 2]
```diff
@@ -209,6 +206,7 @@ export type AusPostResponse = {
 export type UsersResponse = {
     userId: string;
     email: string;
+    createdAt: string;
```
Comment (bitcoinbrisbane) [inline src/interfaces.ts:209]: Not a time based class?

**Review by bitcoinbrisbane — COMMENTED** [review]
See comments

File: src/interfaces.ts (hunk 2 above)
Comment (xiangzhe0812) [reply src/interfaces.ts:209]: `[
    {
        "userId": "09ebefb5-1e2d-4047-8856-0227c9de5ae4",
        "email": "dev@dltx.io",
        "createdAt": "2022-06-27T23:41:31.515Z"
    },
    {
        "userId": "d9abeaf5-50c2-485e-b645-b8557f4d47f4",
        "email": "b6e7543282bbf0627fdc11e649d0be871965fe87b16d0f3ebea39f82ed47ea62",
        "createdAt": "2022-06-30T11:46:01.362Z"
    },
    {
        "userId": "b9625262-0a14-497c-8b07-cd0c4a7d89a9",
        "email": "5b6416265295b815b20e5f8a28c7372a1fa94f3ab30cd6a9944244098eda90ab",
        "createdAt": "2022-07-01T01:06:01.973Z"
    }
]`

I think its good enough

File: src/interfaces.ts [hunk 3]
```diff
@@ -12,12 +12,34 @@ export enum EntityNames {
...
+    Failed = "failed"
+}
+
+export enum RequestType {
+    Signup = "Signup",
+    Verify = "Verify"
+}
+
+//Add more venders
+export enum VenderEnum {
```
Comment (bitcoinbrisbane) [inline src/interfaces.ts:27]:
```suggestion
//Add more vendors
export enum VendorEnum {
```

File: src/services/ThirdPartyService.ts [hunk 4]
```diff
@@ -34,10 +44,77 @@ export class ThirdPartyService implements IThirdPartyService {
...
+            requestBody = {
+                firstName: signupInfo?.firstName,
+                lastName: signupInfo?.lastName,
+                email: signupInfo.email,
+                password: signupInfo.password,
+                referralCode: "",
+                trackAddress: true,
+                createAddress: true
+            };
+        } else if (signupInfo.source === VenderEnum.CoinStash) {
```
Comment (bitcoinbrisbane) [inline src/services/ThirdPartyService.ts:80]:
```suggestion
        } 
        
        if (signupInfo.source === VenderEnum.CoinStash) {
```

File: src/services/ThirdPartyService.ts [hunk 5]
```diff
@@ -34,10 +44,77 @@ export class ThirdPartyService implements IThirdPartyService {
...
+            endPoint = this.config.get(
+                ConfigSettings.COINSTASH_SIGNUP_ENDPOINT
+            );
+            requestBody = {
+                email: signupInfo.email,
+                password: signupInfo.password,
+                displayName: `${signupInfo?.firstName} ${signupInfo?.lastName}`,
+                country: "Australia",
+                token: this.config.get(ConfigSettings.COINSTASH_TOKEN),
+                acceptMarketing: true
```
Comment (bitcoinbrisbane) [inline src/services/ThirdPartyService.ts:90]: Maybe this should be false?

```suggestion
                acceptMarketing: false
```

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Few small things to address

**Review by JasminFarhi — APPROVED** [review]
Looks good to me, probably good to get another sudo dev to look though, just in case.  :)

**Review by amagon270 — DISMISSED** [review]

File: src/interfaces.ts [hunk 6]
```diff
@@ -172,18 +204,17 @@ export type AusPostResponse = {
     // }
 };
 
+export type UsersResponse = {
+    userId: string;
+    email: string;
+    createdAt: Date;
+};
 export type KycResponse = {
```
Comment (bitcoinbrisbane) [inline src/interfaces.ts:212]:
```suggestion
};

export type KycResponse = {
```

File: src/services/ThirdPartyService.ts [hunk 7]
```diff
@@ -34,10 +45,86 @@ export class ThirdPartyService implements IThirdPartyService {
...
     }
+
+    public async signup(
+        signupInfo: UserSignupRequest,
+        ip: string
+    ): Promise<string> {
+        let requestBody = {};
+        let endPoint: string;
+
+        if (signupInfo.source == VendorEnum.GPIB) {
```
Comment (bitcoinbrisbane) [inline src/services/ThirdPartyService.ts:70]: Should this be...

```suggestion
        if (signupInfo.source === VendorEnum.GPIB) {
```

File: src/services/ThirdPartyService.ts [hunk 8]
```diff
@@ -34,10 +45,86 @@ export class ThirdPartyService implements IThirdPartyService {
...
+                firstName: signupInfo?.firstName,
+                lastName: signupInfo?.lastName,
+                email: signupInfo.email,
+                password: signupInfo.password,
+                referralCode: "",
+                trackAddress: true,
+                createAddress: true
+            };
+        }
+        if (signupInfo.source == VendorEnum.CoinStash) {
```
Comment (bitcoinbrisbane) [inline src/services/ThirdPartyService.ts:82]: and here

File: src/services/ThirdPartyService.ts [hunk 9]
```diff
@@ -34,10 +45,86 @@ export class ThirdPartyService implements IThirdPartyService {
...
+                email: signupInfo.email,
+                password: signupInfo.password,
+                displayName: `${signupInfo?.firstName} ${signupInfo?.lastName}`,
+                country: "Australia",
+                token: this.config.get(ConfigSettings.COINSTASH_TOKEN),
+                acceptMarketing: false
+            };
+        }
+
+        if (signupInfo.source == VendorEnum.EasyCrypto) {
```
Comment (bitcoinbrisbane) [inline src/services/ThirdPartyService.ts:96]: and here

**Review by bitcoinbrisbane — COMMENTED** [review]
Few minor things to check.

**Review by bitcoinbrisbane — APPROVED** [review]

**Review by amagon270 — APPROVED** [review]

**Pull request merged**